Document openssl thread-safety

richsalz opened the pull request:
Now that we have a man7, it's about time that OpenSSL had a manpage like this :)
Could be backported to 1.1.1

t-j-h commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,58 @@
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is ofen an I<object>_up_ref() API that can be used to increase the
```
often

richsalz replied:
fixup commit to fix typo fixed, thanks!

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,66 @@
+To repeat: I<most objects are not safe for simultaneous use>.
+Exceptions to this should be documented on the specific manual pages, and
+some general high-level guidance is given here.
+
```
I think there should be some discussion about calling functions from multiple threads, but using *different* objects, i.e. in most cases it is safe to call the same function in multiple threads as long as you use a different object.

There may be some exceptions to this rule, e.g. in some cases where global state is modified (OBJ_new_nid() and OBJ_create() spring to mind as instances which are not thread safe at all (this might actually be considered a bug for these functions))

richsalz replied:
I added a couple of paragraphs in the new commit.

kroeckx commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,66 @@
+modifier.
+Another instance of safe use is when an object's internal state,
+such as cached values, are done with locks.
+One example of this is the reference counting API's described above.
```
Should we be more explicit and say that const doesn't mean immutable or physical constant, but does mean logical constant and thread safe? It's then up to the object itself to take a lock if it wants to modify something, but remain logically the same. For instance X509_cmp() currently takes 2 `const X509 *`, but will internally remove the const to update cached values. It will internally take a lock to update that cache, and so it's actually thread safe, and logically constant. X509_check_purpose() has a comment that it also wants const but can't because it wants to update the cache, and so could actually also have the const.

I would remove the "mutable const" part, it's not really relevant.

richsalz replied:
I tried to add some text about this, in the commit I am about to push.

kroeckx replied:
The "Another instance", is really the same as the X509_cmp() example, cache something.

Note that none of the up_ref() calls currently take a const because they modify the counter, but as documented here, they could and probably should.

richsalz replied:
I couldn't think of another type of thing.  But it's "another instance" because it might not be obvious to developers.  If you can think of something that is different, let me know.  If you think I should remove that other instance (I would rather not because of the obviousness I mentioned), let me knwo.

I created #13796 for the "const" in upref/downref.

mattcaswell replied:
How is a user supposed to use the information provided in this paragraph? Really, I think you are saying that there are some functions that mutate objects that may still be thread safe. We should seek to document what those functions are (doesn't have to be in this PR)

richsalz replied:
It's a warning that the behavior, if not otherwise documented, is undefined.  "Here be dragons."

mattcaswell commented:
The CI failure is relevant.

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,98 @@
+some general high-level guidance is given here.
+
+One major use of the OpenSSL thread API is to implement reference counting.
+Many object within OpenSSL are reference-counted, so resources are not
```
Typo: objects

richsalz replied:
fixed typo.

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,98 @@
+
+The various C<set1> and C<get1> functions associated with an object are
+used to set or get attributes of an object; they increase the refernece
+count.
```
This seems quite confusing. They don't increase the ref count of the object. They *may* increase the ref count of the thing being set into or got from the object if the thing is a ref countable object. 

Also, typo: reference.

I find this whole discussion about ref counting useful, but misplaced on a page about threads. This is only very tangentially related to threads. Ref counting may be used as a mechanism to share a single object across multiple threads - but there are plenty of other scenarios where ref counting is useful.

richsalz replied:
Fixed typo.  Added "increase the reference count if it's an object that has them."

It might be misplaced -- I disagree, reference counts and ownership are related to multi-threaded programming -- but I think it's important to explain somewhere. Perhaps a future edit will create an openssl-refcount page, but for now I think it should stay.

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,98 @@
+mutate an object, such as setting elements of a private or public key,
+while another thread is using that object, such as verifying a signature.
+
+The same API's can often be used simultaneously on different objects
```
s/often/usually/  ?

richsalz replied:
I said "can often be used" because I did not feel comfortable making the further claim of "usually."  I don't really want to change it, but will if you think it's important.

mattcaswell replied:
I do think its important. In the vast majority of cases this is true, and the preferred way of ensuring thread safety.

richsalz replied:
Okay changed "can often be used" to "usually be used"

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,98 @@
+when other objects have been created from that.
+For example, an B<SSL_CTX> object created by L<SSL_CTX_new(3)> is used
+to create per-connection B<SSL> objects by calling L<SSL_new(3)>.
+This is not safe, no matter if multiple threads are involved or not.
```
I found this confusing. I think you're trying to say that modifying factory objects after objects have already been created is unsafe. But it reads like you are saying creating per-connection SSL objects from an SSL_CTX is unsafe.

richsalz replied:
Yes, "modifying factory objects" is the concept I was looking for, thanks.

richsalz commented:
fixup commit pushed, including (AARGH CAN'T BELIEVE I DID THAT), the doc-nits failure.

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,100 @@
+=head1 DESCRIPTION
+
+In this man page, we use the term B<thread-safe> to indicate that an
+object can be used by multiple threads at the same time.
```
I'd rather say "indicate that an object or function can be used". Thread safety is about more than just objects.

richsalz replied:
added "object *or function* can be used"

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,100 @@
+used to set or get attributes of an object; they increase the reference
+count of the attribute, if the attribute is an object that has them.
+The C<set0> and C<get0> functions are similar, but do not increase the
+reference count, passing "ownership" appropriately.
```
I think this is not entirely clear. A "set0" passes ownership from the caller to the object. However with a "get0" ownership remains with the object.

With a "set1" ownership remains with the caller, and with a "get1" ownership is passed from the object to the caller.

richsalz replied:
Thanks for that input, i used it to rewrite like this:
```
Many objects have set and get API's to set attributes in the object. A C<set0> passes ownership
from the caller to the object and a  C<get0> returns a pointer but the attribute ownership remains
with the object. A C<set1> or C<get1> function does not change the ownership, and it is up to the
caller to free the object or resource. Functions with a simple C<set> or C<get> are historical; and
the documentation must state how the references are handled. Get methods are often thread-safe
as long as the ownership requirements are met. Set methods are generally not thread-safe as discussed below.
```

kroeckx replied:
I don't know if it's always the case, but a get1 and set1 increases the reference count, and so a for get1 the application should free it, for a set1 it will be freed as part of freeing the struct. I guess an other implementation of a get1 and set1 can be that it duplicates it, but it should always free it. So either the ownership is moved or shared.

richsalz replied:
Sorry, I don't understand what you're trying to say.  That's my fault.  Can you propose specific wording changes to the text quoted in the comment above?

kroeckx replied:
You say `A C<set1> or C<get1> function does not change the ownership`, but it does get1 does.

richsalz replied:
Thanks!  How about "set1 transfers ownership of the attribute to the object and get1 returns a copy" ?

kroeckx replied:
I guess I edited my previous reply before you saw it.

kroeckx replied:
The `1` functions always make a copy, which can be implemented as a reference count increment. So on get, it becomes a (shared) owner, on set1 it stays a (shared) owner.

richsalz replied:
> The `1` functions always make a copy, which can be implemented as a reference count increment. So on get, it becomes a (shared) owner, on set1 it stays a (shared) owner.

Manipulating the refcount isn't a copy tho.  It's `foo->p = p` versus `foo->p = strdup(p);`

But anyway how about this:
```
Many objects have set and get API's to set attributes in the object.
A C<set0> passes ownership from the caller to the object and a
C<get0> returns a pointer but the attribute ownership
remains with the object and a copy is returned.
A C<set1> or C<get1> function does not change the ownership, but instead
updates the attribute's reference count so that the object is shared 
between the caller and the object; the caller must free the attribute when
finished.
Functions with a simple C<set> or C<get> are historical; and the documentation
must state how the references are handled.
Get methods are often thread-safe as long as the ownership requirements are
met and shared objects are not modified.
Set methods, or modifying shared objects, are generally not thread-safe 
as discussed below.
```

t8m replied:
`s/remains with the object and a copy is returned/remains with the object and just a reference to it is returned`

Are there any cases where a full dup operation is done on set1/get1? Not that it matters much as it is always better to treat them as shared reference just with an increased refcount.

richsalz replied:
> Are there any cases where a full dup operation is done on set1/get1?

strdup kinds of things.  but sure, your reasoning makes sense; fixup commit pushed.

mattcaswell commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,100 @@
+Note that a C<const> declaration does not mean immutable; for example
+L<X509_cmp(3)> takes pointers to C<const> objects, but the implementation
+uses a C cast to remove that so it can lock objects, generate and cache
+a DER encoding,, and so on.
```
double comma here

mattcaswell replied:
I don't think that this last sentence adds much value and just confuses the matter. The example given  is really just an implementation detail that should perhaps be noted in the man page for the function. IIUC, it doesn't impact the thread safety in any way.

richsalz replied:
I think for the less-experienced C programmers this is useful info.

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+
+In particular, being configured for threads support does not imply that
+all OpenSSL objects are thread-safe.
+To repeat: I<most objects are not safe for simultaneous use>.
```
`To emphasize:` would sound less parenting to me.

richsalz replied:
Changed to "emphasize"

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is often an I<object>_up_ref() API that can be used to increase the
```
s/API/API call/ possibly?

richsalz replied:
I just changed "API" to "function" to avoid the issue. :)

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+remains with the object and a reference to it is returned.
+A C<set1> or C<get1> function does not change the ownership, but instead
+updates the attribute's reference count so that the object is shared
+between the caller and the object; the caller must free the attribute when
```
`free the returned attribute`

richsalz replied:
Changed to "caller must free the returned attribute"

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+updates the attribute's reference count so that the object is shared
+between the caller and the object; the caller must free the attribute when
+finished.
+Functions with a simple C<set> or C<get> are historical; and the documentation
```
Sometimes they are not historical if the attribute is a simple scalar and not object. So perhaps: `Functions that provide access to ref-counted attributes with a simple ....`

richsalz replied:
Changed to
```
Functions that involve attributes that have reference counts themselves,
but are named with just C<set> or C<get> are historical; and the documentation
must state how the references are handled.
```

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+Objects are thread-safe
+as long as the API's being invoked don't modify the object; in this
+case the parameter is usually marked in the API as C<const>.
+Not all parameters are marked this way.
```
`... this way for historical API compatibility reasons.`

richsalz replied:
I don't think so, because the project has already agreed that adding `const` to parameters is okay and compatible.  It's more that nobody changed the code to add casts as needed.

t8m replied:
In some cases like function pointer types the historical API compatibility is the reason. But OK, I do not insist on that.

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,104 @@
+uses a C cast to remove that so it can lock objects, generate and cache
+a DER encoding, and so on.
+
+Another instance of thread-safety is when an object's internal state,
```
s/when an/when updates to an/

richsalz replied:
Changed to "when updates to an object's internal state"

t8m reviewed (commented):
Just a few nits, otherwise good work!

t8m commented on doc/man7/openssl-threads.pod:
```diff
@@ -55,8 +56,8 @@ L<X509_cmp(3)> takes pointers to C<const> objects, but the implementation
 
-Another instance of thread-safety is when an object's internal state,
-such as cached values, are done with locks.
+Another instance of thread-safety is when an updates to an object's
```
s/an updates/updates/

richsalz replied:
typo fixed, thanks for the careful reading.

t8m reviewed (changes requested):
Just one typo left.

kroeckx commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,105 @@
+The same API's can usually be used simultaneously on different objects
+without interference.
+For example, two threads can calculate a signature using two different
+B<EVP_PKEY> objects.
```
EVP_PKEY_CTX? I expect the same EVP_PKEY can be used by multiple threads, but looking at the code, it seems that at least in master that's not the case.

richsalz replied:
What code were you looking at?  Just curious because I had the same expectation about `EVP_PKEY` being threadsafe.  For now, just changed it to `EVP_PKEY_CTX` for interest of getting this done

kroeckx replied:
EVP_PKEY now has an `operation_cache` and `cache` it didn't have in 1.1.1. We seem to be using at least the `operation_cache` without a lock, and it looks to me like it should use a lock. The `cache` is probably fine.

richsalz replied:
You should create an issue out of your comment.  But I did use the CTX object so this is resolved.

richsalz replied:
I created #13818 so that I can mark this conversation as resolved.

t8m commented:
The PR now seems to include commits that are already on master branch. Could you please do a clean rebase so there are only the relevant commits included?

richsalz commented:
>please do a clean rebase

better now?

t8m commented:
There are still the commits from PR #13742 included.

richsalz commented:
"This time for sure."

t8m commented on doc/man3/CRYPTO_THREAD_run_once.pod:
```diff
@@ -113,12 +113,13 @@ The other functions return 1 on success, or 0 on error.
 On Windows platforms the CRYPTO_THREAD_* types and functions in the
-openssl/crypto.h header are dependent on some of the types customarily
-made available by including windows.h. The application developer is
+F<openssl/crypto.h> header are dependent on some of the types customarily
```
Elsewhere @levitte used `F<< <openssl/crypto.h> >>`. Could you please do the same here and in other places mentioning headers below.

richsalz replied:
I had deliberately not done `F<< <` to avoid stepping on @levitte's PR.  Changed.

t8m reviewed (commented):
Just one more nit, otherwise LGTM.

levitte commented on doc/man7/openssl-threads.pod:
```diff
@@ -0,0 +1,105 @@
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is often an I<object>_up_ref() function that can be used to increase
```
Unfortunately, I don't think that pod2xxx are smart enough to format function names with markup in a sane way, so it needs a bit of help.  `B<I<object>_up_ref>()`

Function names *without* any pod markup (i.e. something like `BLAH_foo()`) in them are fine, pod2xxx will apply a result that makes them appropriately bold.

richsalz commented:
Since there was a duplication of effort in the run-once manpage, I squashed and rebased.  Thanks all for the feedback.
@t8m no changes to openssl-threads page.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to master. Thank you for the contribution!
